File: src/llm.js
const DEFAULT_OLLAMA_BASE_URL = "http://127.0.0.1:11434";

function localFallbackResponse(prompt) {
  const cleaned = String(prompt || "").trim();
  if (!cleaned) {
    return "Please send a message so I can help.";
  }
  return `(local fallback) I received: '${cleaned}'.\nRun Ollama and set AI_BACKEND=ollama for live model responses.`;
}

function getBackend() {
  return (process.env.AI_BACKEND || "ollama").trim().toLowerCase();
}

async function ollamaGenerate(prompt) {
  const baseUrl = (process.env.OLLAMA_BASE_URL || DEFAULT_OLLAMA_BASE_URL).replace(/\/$/, "");
  const model = (process.env.OLLAMA_MODEL || "llama3.2").trim();

  const response = await fetch(`${baseUrl}/api/generate`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      model,
      prompt,
      system: "You are a concise and helpful AI assistant.",
      stream: false
    })
  });

  if (!response.ok) {
    const details = await response.text();
    throw new Error(`Ollama request failed (${response.status}): ${details}`);
  }

  const data = await response.json();
  const content = data?.response;
  return (content || "I could not generate a response.").trim();
}

export function getRuntimeMode() {
  return getBackend() === "ollama" ? "ollama" : "fallback";
}

export async function generateResponse(prompt) {
  if (getBackend() !== "ollama") {
    return localFallbackResponse(prompt);
  }

  try {
    return await ollamaGenerate(prompt);
  } catch {
    return localFallbackResponse(prompt);
  }
}

File: ai_server/llm.py
from __future__ import annotations

import json
import os
from dataclasses import dataclass
from urllib import request


@dataclass
class LLMConfig:
    model: str = "gpt-4o-mini"
    system_prompt: str = "You are a concise and helpful AI assistant."


def _local_fallback_response(user_prompt: str) -> str:
    cleaned = user_prompt.strip()
    if not cleaned:
        return "Please send a message so I can help."

    return (
        "(local fallback) I received: "
        f"'{cleaned}'.\n"
        "Set OPENAI_API_KEY to enable live model responses."
    )


def _openai_chat_completion(prompt: str, cfg: LLMConfig, api_key: str) -> str:
    payload = {
        "model": cfg.model,
        "messages": [
            {"role": "system", "content": cfg.system_prompt},
            {"role": "user", "content": prompt},
        ],
        "temperature": 0.7,
    }

    req = request.Request(
        "https://api.openai.com/v1/chat/completions",
        data=json.dumps(payload).encode("utf-8"),
        headers={
            "Content-Type": "application/json",
            "Authorization": f"Bearer {api_key}",
        },
        method="POST",
    )

    with request.urlopen(req, timeout=30) as response:
        body = json.loads(response.read().decode("utf-8"))

    content = body["choices"][0]["message"]["content"]
    return (content or "").strip() or "I could not generate a response."


def generate_response(user_prompt: str, config: LLMConfig | None = None) -> str:
    cfg = config or LLMConfig()
    api_key = os.getenv("OPENAI_API_KEY", "").strip()

    if not api_key:
        return _local_fallback_response(user_prompt)

    return _openai_chat_completion(user_prompt, cfg, api_key)
